File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& w) {
        int l=w[0].length();
        int x=w.size();
        int winsize=l*x;
        vector<int> ans;
        unordered_map<string,int> f;
        for(auto i:w){
            f[i]++;
        }
        for(int i=0;i<l;i++){
            unordered_map<string,int> seen;            
            int j=i;    // left ptr
            int c=0;    // count of matched words     
            for(int k=i;k+l<=s.size();k+=l){ //k is right ptr
                string a=s.substr(k,l);                
                if(f.find(a)!=f.end()){
                    seen[a]++;
                    c++;
                    while(seen[a]>f[a]){
                        string lw=s.substr(j,l);
                        seen[lw]--;
                        c--;
                        j+=l;
                    }
                    if(c==x) ans.push_back(j);
                }else{
                    seen.clear();
                    c=0;
                    j=k+l;
                }
            }
        }
        return ans;
    }
};

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        ans=[]
        if not words or not s:
            return ans
        wl = len(words[0]) # length of each word
        l = len(words) # no. of words
        f={} # dictinory to store words 
        for i in words:
            f[i]= f.get(i,0)+1

        for i in range(wl):
            seen={}
            left,count=i,0
            for right in range(i,len(s)-wl+1,wl):
                w=s[right:right+wl]
                if w in f:
                    seen[w]= seen.get(w,0)+1                  
                    count+=1
                    while seen[w]>f[w]:
                        a=s[left:left+wl]
                        seen[a]-=1
                        count-=1
                        left+=wl
                    if count==l:
                        ans.append(left)
                else:
                    seen.clear()
                    count=0
                    left=right+wl
        return ans
